### DISCUSSION TITLE
Running Rook/Ceph with SELinux Policies, not with Privileged Pods

### DISCUSSION THREAD

**Discussion by KKonak** [rook/rook#13546] [Q&A] [answered]
I am trying to run Rook/Ceph with SELinux enforcing, I was successful setting the parameter in the helm chart "hostpathRequiresPrivileged" to get it working. But I am wanting to use it with a SELinux policy that restricts Rook/Ceph access to just what it needs to function. 

I am not well versed in creating SELinux policies so I have been exploring different policy generators, but I'm wondering if Rook provides an SELinux policy that I couldnt find in the repo. Or if someone has had success in creating a policy I would appreciate some assistance.

**Comment by travisn** [comment]
There is not a more specific policy for the same.

**Reply by KKonak** [reply]
Running ceph pods as privleged isnt exactly good security posture. Are there any plans to provide SELinux policies in the future?
Perhaps you could explain what "hostpathRequiresPrivleged" does under the hood? Is it giving privilege to just access the host path or to all pod processes? Thank you for the response.

**Reply by travisn** [reply]
Are you running in OpenShift or another environment with SELinux enabled? For OpenShift, we have the [SecurityContextConstraints](https://github.com/rook/rook/blob/master/deploy/examples/operator-openshift.yaml#L1-L99). One of the core needs for privileged access is access to `/dev`, and the data stored to the host path for mons. We are open to contributions in this area if you have specific suggestions.

**Reply by KKonak** [reply]
Thanks for the information, I am running an environment with SELinux enabled as a requirement. I have created a working starter policy that applies to container_t context. Now I want to label rook/ceph pods with its own custom source context (so my policies arent applied to all pods with container_t). 

In the [rook-ceph-values.yaml](https://github.com/rook/rook/blob/master/deploy/charts/rook-ceph/values.yaml#L69) "containerSecurityContext" is an option which I presumed would have the[ k8s native "seLinuxOptions"](https://kubernetes.io/docs/tasks/configure-pod-container/security-context/#assign-selinux-labels-to-a-container) field to label the container with a source context. Would this be a reasonable addition to add to the helm charts to make labeling easier so it can be configured for seLinux environments?

**Reply by travisn** [reply, marked as answer]
The `containerSecurityContext` only applies to the operator and toolbox pods, but at least it should already support the full securityContext options such as `seLinuxOptions`. 

For all the ceph pods, the only security context currently applied is found in [these helper methods](https://github.com/rook/rook/blob/bc78b6cad21c62da1501dd19b04d954a51079afe/pkg/operator/ceph/controller/spec.go#L727-L758), which don't have a way to override from the CR settings. Seems like we need a `seLinuxOptions` setting in the CephCluster CR to allow setting this (and the helm chart would also inherit this setting).

**Reply by KKonak** [reply]
Okay I've gathered that thank you! Im currently working on patching those helper methods in controller/spec.go so its good to know im on the right track to labeling them properly. Am I correct in seeing that controller deploys mons and mgr then the [osd spec.go](https://github.com/rook/rook/blob/master/pkg/operator/ceph/cluster/osd/spec.go#L913) handles labeling those pods here?

If rook could add an seLinuxOptions setting that would be a great addition for the future, so in future versions we dont have to look through and repatch the go files.

Thank you for your responses, this dialogue has been a great help.

**Reply by travisn** [reply]
The OSD pods have several init containers, as you see that one specifically requests a different context.

**Reply by KKonak** [reply]
For documentation and to help others, can confirm success with patching helper methods like so:
```
func PodSecurityContext() *v1.SecurityContext {
    privileged := HostPathRequiresPrivileged()

    return &v1.SecurityContext{
        Privileged: &privileged,
        SELinuxOptions: &v1.SELinuxOptions{
            Type: "rook_ceph_t",
        },
    }
}
```
The ceph pods seem to deploy now with the custom context.
